### PR TITLE
feat: add `PostprocessingEvaluator`

### PR DESCRIPTION
**Pull request by iajoiner**
# Rationale for this change
For unprovable expressions it is necessary to evaluate them in the verifier. Hence it is necessary to add an evaluation framework to evaluate any `Expression`.

# What changes are included in this PR?
- add `PostprocessingEvaluator`
- add division utils

# Are these changes tested?
Pending

### CODE REVIEW

File: crates/proof-of-sql/src/base/database/owned_column.rs [hunk 1]
```diff
@@ -56,6 +56,23 @@ impl<S: Scalar> OwnedColumn<S> {
...
+            Self::SmallInt(col) => S::from(col[index]),
+            Self::Int(col) => S::from(col[index]),
+            Self::BigInt(col) => S::from(col[index]),
+            Self::Int128(col) => S::from(col[index]),
+            Self::Scalar(col) => col[index],
+            Self::Decimal75(_, _, col) => col[index],
+            Self::VarChar(col) => S::from(col[index].clone()),
+            Self::TimestampTZ(_, _, col) => S::from(col[index]),
+        })
+    }
```
Comment (drcapybara) [inline crates/proof-of-sql/src/base/database/owned_column.rs:74]: Would it be better to error here instead of returning ```Option```? Index OOB seems like it could be a critical failure path

File: crates/proof-of-sql/src/base/math/decimal.rs [hunk 2]
```diff
@@ -42,13 +31,17 @@ pub enum DecimalError {
...
-impl From<IntermediateDecimalError> for ConversionError {
-    fn from(err: IntermediateDecimalError) -> ConversionError {
-        DecimalConversionError(IntermediateDecimalConversionError(err))
+/// Result type for decimal operations.
+pub type DecimalResult<T> = Result<T, DecimalError>;
+
+// This exists because `TryFrom<arrow::datatypes::DataType>` for `ColumnType` error is String
+impl From<DecimalError> for String {
+    fn from(error: DecimalError) -> Self {
+        error.to_string()
```
Comment (drcapybara) [inline crates/proof-of-sql/src/base/math/decimal.rs:44]: Since we're here right now, it might be prudent to actually go implement the error variant `ColumnType` error

File: crates/proof-of-sql/src/sql/ast/numerical_util.rs [hunk 3]
```diff
@@ -121,6 +162,55 @@ pub(crate) fn add_subtract_columns<'a, S: Scalar>(
...
+                .precision_value()
+                .expect("Precision of a valid column type should be valid"),
+        )
+        .expect("Precision of a valid column type should be valid"),
+        res_type
+            .scale()
+            .expect("Can't add or subtract non-numeric types"),
+        res,
+    )
+}
```
Comment (drcapybara) [inline crates/proof-of-sql/src/sql/ast/numerical_util.rs:212]: Why is add and subtract together? Would it make more sense to have separate functions for add and subtract?

File: crates/proof-of-sql/src/sql/postprocessing/evaluate.rs [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+        }
+    }
+    /// Creates a new `PostprocessingEvaluator` with the given column mapping and within aggregation scope.
+    #[allow(dead_code)]
+    pub(crate) fn new_agg(table: OwnedTable<S>) -> Self {
+        Self {
+            table,
+            _in_agg_scope: true,
+        }
+    }
```
Comment (drcapybara) [inline crates/proof-of-sql/src/sql/postprocessing/evaluate.rs:41]: is there a good reason to allow unused functions and fields here? If they will end up being used in a future PR then it seems fine

File: crates/proof-of-sql/src/sql/postprocessing/evaluate.rs [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+    ) -> PostprocessingResult<OwnedColumn<S>> {
+        let left = self.visit_expr(left)?;
+        let right = self.visit_expr(right)?;
+        let length = left.len();
+        assert_eq!(length, right.len());
+        check_dtypes(left.column_type(), right.column_type(), op)?;
+        match op {
+            BinaryOperator::And => Ok(OwnedColumn::Boolean(
+                left.as_boolean()
+                    .unwrap()
```
Comment (drcapybara) [inline crates/proof-of-sql/src/sql/postprocessing/evaluate.rs:135]: this unwrap seems like it should be handled

File: crates/proof-of-sql/src/sql/postprocessing/evaluate.rs [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+            BinaryOperator::Equal
+            | BinaryOperator::GreaterThanOrEqual
+            | BinaryOperator::LessThanOrEqual => {
+                let lhs_scale = left.column_type().scale().unwrap_or(0);
+                let rhs_scale = right.column_type().scale().unwrap_or(0);
+                let max_scale = lhs_scale.max(rhs_scale);
+                let lhs_upscale_factor =
+                    scale_scalar(S::ONE, max_scale - lhs_scale).expect("Invalid scale factor");
+                let rhs_upscale_factor =
+                    scale_scalar(S::ONE, max_scale - rhs_scale).expect("Invalid scale factor");
```
Comment (drcapybara) [inline crates/proof-of-sql/src/sql/postprocessing/evaluate.rs:158]: these ```.expects``` seem like they are reporting a failure, but it seems like expect should be used when we are 100% certain that this cannot fail, so a message explaining why the failure cannot occur seems more appropriate

File: crates/proof-of-sql/src/sql/postprocessing/evaluate.rs [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+                match op {
+                    BinaryOperator::Equal => {
+                        let res: Vec<bool> = (0..length)
+                            .map(|i| {
+                                left.scalar_at(i).unwrap() * lhs_upscale_factor
+                                    == right.scalar_at(i).unwrap() * rhs_upscale_factor
```
Comment (drcapybara) [inline crates/proof-of-sql/src/sql/postprocessing/evaluate.rs:164]: this unwrap seems like it should be handled, if this cannot panic then should it be .expect?

File: crates/proof-of-sql/src/sql/postprocessing/evaluate.rs [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+                                lhs.signed_cmp(&rhs) != Ordering::Less
+                            })
+                            .collect();
+                        Ok(OwnedColumn::Boolean(res))
+                    }
+                    BinaryOperator::LessThanOrEqual => {
+                        let res: Vec<bool> = (0..length)
+                            .map(|i| {
+                                let lhs = left.scalar_at(i).unwrap() * lhs_upscale_factor;
+                                let rhs = right.scalar_at(i).unwrap() * rhs_upscale_factor;
```
Comment (drcapybara) [inline crates/proof-of-sql/src/sql/postprocessing/evaluate.rs:183]: this unwrap seems like it should be handled, if this cannot panic then should it be .expect?

File: crates/proof-of-sql/src/sql/postprocessing/evaluate.rs [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+                if precision.value() > 38 {
+                    return None;
+                }
+            }
+            if let ColumnType::Decimal75(precision, _) = right_dtype {
+                if precision.value() > 38 {
+                    return None;
+                }
+            }
+            // TODO: inequality support for timestamps
```
Comment (drcapybara) [inline crates/proof-of-sql/src/sql/postprocessing/evaluate.rs:237]: timestamp inequality support was merged in #52, is this comment referring to something different?

File: crates/proof-of-sql/src/sql/ast/numerical_util.rs (hunk 3 above)
Comment (iajoiner) [reply crates/proof-of-sql/src/sql/ast/numerical_util.rs:212]: @drcapybara They are together since add and subtract are essentially the same other than a sign.

File: crates/proof-of-sql/src/sql/postprocessing/evaluate.rs (hunk 4 above)
Comment (iajoiner) [reply crates/proof-of-sql/src/sql/postprocessing/evaluate.rs:41]: Yes. I will add aggregation once `SelectExpr` is added in the next PR

File: crates/proof-of-sql/src/sql/postprocessing/evaluate.rs [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+impl<S: Scalar> PostprocessingEvaluator<S> {
+    /// Creates a new `PostprocessingEvaluator` with the given owned table and group by columns.
+    pub fn new(table: &OwnedTable<S>, group_by_columns: &[Identifier]) -> Self {
+        Self {
+            table: table.clone(),
+            _group_by_columns: group_by_columns.to_vec(),
+        }
+    }
+    /// Builds a `proofs::sql::ast::ProvableExprPlan` from a `proof_of_sql_parser::intermediate_ast::Expression`
+    pub fn build(&self, expr: &Expression) -> PostprocessingResult<OwnedColumn<S>> {
```
Comment (JayWhite2357) [inline crates/proof-of-sql/src/sql/postprocessing/evaluate.rs:35]: `build` seems like the wrong name.
Also, the docs appear to be wrong.

File: crates/proof-of-sql/src/sql/postprocessing/evaluate.rs [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+use proof_of_sql_parser::{
+    intermediate_ast::{BinaryOperator, Expression, Literal, UnaryOperator},
+    Identifier,
+};
+
+/// Evaluator that evaluates a `proof_of_sql_parser::intermediate_ast::Expression`
+/// on an `OwnedTable` and returns an `OwnedColumn`.
+pub struct PostprocessingEvaluator<S: Scalar> {
+    table: OwnedTable<S>,
+    _group_by_columns: Vec<Identifier>,
```
Comment (JayWhite2357) [inline crates/proof-of-sql/src/sql/postprocessing/evaluate.rs:23]: Why not just add this in a later PR rather than making it unused right now?

File: crates/proof-of-sql/src/base/database/scalar_and_i256_conversions.rs [hunk 12]
```diff
@@ -14,7 +14,7 @@ const MAX_SUPPORTED_I256: i256 = i256::from_parts(
 pub fn convert_scalar_to_i256<S: Scalar>(val: &S) -> i256 {
     let is_negative = val > &S::MAX_SIGNED;
     let abs_scalar = if is_negative { -*val } else { *val };
-    let limbs = abs_scalar.into();
+    let limbs: [u64; 4] = abs_scalar.into();
```
Comment (JayWhite2357) [inline crates/proof-of-sql/src/base/database/scalar_and_i256_conversions.rs:17]: Is this necessary? No reason to include it in the PR if not.

File: crates/proof-of-sql/src/sql/ast/numerical_util.rs [hunk 13]
```diff
@@ -121,6 +163,44 @@ pub(crate) fn add_subtract_columns<'a, S: Scalar>(
...
+            if is_subtract {
+                lhs.scalar_at(i).unwrap() * lhs_upscale_factor
+                    - rhs.scalar_at(i).unwrap() * rhs_upscale_factor
+            } else {
+                lhs.scalar_at(i).unwrap() * lhs_upscale_factor
+                    + rhs.scalar_at(i).unwrap() * rhs_upscale_factor
+            }
+        })
+        .collect();
+    OwnedColumn::from_scalars(res, res_type).expect("Invalid column type")
```
Comment (JayWhite2357) [inline crates/proof-of-sql/src/sql/ast/numerical_util.rs:201]: I really don't like converting to Scalars, subtracting, and then converting back.
Additionally, scaling isn't always needed.
For instance, when subtracting two `BigInt` columns, we should do subtraction of `i64`s. Scaling is unneeded in that case.

File: crates/proof-of-sql/src/sql/ast/numerical_util.rs [hunk 14]
```diff
@@ -121,6 +163,44 @@ pub(crate) fn add_subtract_columns<'a, S: Scalar>(
     res
 }
 
+// Add or subtract two owned columns together.
+pub(crate) fn add_subtract_owned_columns<S: Scalar>(
```
Comment (JayWhite2357) [inline crates/proof-of-sql/src/sql/ast/numerical_util.rs:167]: Can you put these methods in a separate PR along with unit tests for them?

Comment (JayWhite2357) [reply crates/proof-of-sql/src/sql/ast/numerical_util.rs:167]: Additionally, why not just `impl Sub for &OwnedColumn<S>`?

**Review by JayWhite2357 — CHANGES_REQUESTED** [review]

File: crates/proof-of-sql/src/base/database/scalar_and_i256_conversions.rs (hunk 12 above)
Comment (iajoiner) [reply crates/proof-of-sql/src/base/database/scalar_and_i256_conversions.rs:17]: I think so since I added `Into<BigInt>` to `Scalar` definition

File: crates/proof-of-sql/src/base/database/owned_column.rs [hunk 15]
```diff
@@ -56,6 +56,60 @@ impl<S: Scalar> OwnedColumn<S> {
...
+            ColumnType::Scalar => Some(OwnedColumn::Scalar(scalars.to_vec())),
+            ColumnType::Decimal75(precision, scale) => {
+                Some(OwnedColumn::Decimal75(precision, scale, scalars.to_vec()))
+            }
+            ColumnType::TimestampTZ(tu, tz) => scalars
+                .iter()
+                .map(|s| -> Option<i64> { (*s).try_into().ok() })
+                .collect::<Option<Vec<i64>>>()
+                .map(|s| OwnedColumn::TimestampTZ(tu, tz, s)),
+            // Boolean and VarChar columns are not supported by this function
```
[the remark this answers is not in the document]
Comment (iajoiner) [reply crates/proof-of-sql/src/base/database/owned_column.rs:108]: Of course we can support it. It's just that there isn't any need now.

**Comment by iajoiner** [comment]
Closed in favor of #81.
